File: c/algorithms/quick_select/quick_select.c
/******************************************************************************
 * File: quick_select.c
 * Author: Patrick Payne
 * Date Created: Oct 21, 2013
 * Purpose: Implements a basic array quick select algorithm.
 * Copyright 2013 by Patrick Payne.
 *****************************************************************************/

#include "quick_select.h"

#include <assert.h>
#include <stdlib.h>
#include <stdbool.h>

/*! @brief Given an array of integers, determine the kth smallest.
 *     element.
 *  @param list A pointer to the first element in the array.
 *  @param size The number of valid elements in the array.
 *  @param k One more than the number of elements smaller than
 *     the number returned.
 *  @return The kth smallest integer in list.
 */
int QuickSelect(int *list, int size, int k) {
    // Check that the inputs are valid
    assert(list != NULL);
    assert(size > 0);
    assert(k > 0);
    assert(size >= k);

     int left = 0;
    int right = size - 1;

     while (true) {
        // Choose a pivot, and partition the array.
        int pivot = left;
        pivot = Partition(list, left, right, pivot);

        // Determine how many elements are smaller or equal to
        // the pivot element.
        int pivot_position = pivot + 1;
        if (pivot_position > k) {
            // There are more than k-1 smaller elements.
            right = pivot - 1;
        } else if (pivot_position < k) {
            // There are fewer than k-1 smaller elements.
            left = pivot + 1;
         } else {
            // We have found the kth item.
            return list[pivot];
        }
    }
}


/*! @brief Partitions an array of ints into elements smaller than the
 *    pivot element and elements larger than the pivot, in place.
 *  @param list A pointer to the first element in the array.
 *  @param left The starting index of the subarray.
 *  @param right The ending index of the subarray.
 *  @param pivot The index of the pivot element.
 *  @return The new index of the pivot.
 */
int Partition(int *list, int left, int right, int pivot) {
    int pivot_value = list[pivot];
    Swap(list, pivot, right);

    // Keep all elements smaller than the pivot to the left, and all
    // larger and equal items to the right.
    int store_index = left;
    for (int i = left; i < right; i++) {
        if (list[i] < pivot_value) {
            Swap(list, store_index, i);
            store_index++;
        }
    }

    // Move the pivot back to its proper location, and return it.
    // if ((store_index != left)) {
      Swap(list, right, store_index);
    // }
    return store_index;
}


/*! @brief Swaps two elements in an array of integers. */
void Swap(int *list, int index1, int index2) {
    int temp = list[index1];
    list[index1] = list[index2];
    list[index2] = temp;
}

File: python/algorithms/quickselect/quickselect.py
#!/usr/bin/env python3
import random
import unittest


def swap(in_list, index1, index2):
    '''Given a list in_list, swap elements at index1 and index2.'''
    temp = in_list[index1]
    in_list[index1] = in_list[index2]
    in_list[index2] = temp


def partition(in_list, start, end, pivot_index):
    '''Partition in_list[start:end+1] around the item at pivot_index, in place.

    Returns: The final index of the pivot.
    '''
    assert start <= pivot_index <= end

    # Stick the pivot at the end to make partitioning easier.
    pivot = in_list[pivot_index]
    swap(in_list, pivot_index, end)

    # Keep all smaller items to the left, larger items of partition_index.
    partition_index = start
    for index in range(start, end):
        if in_list[index] <= pivot:
            swap(in_list, partition_index, index)
            partition_index += 1

    # Put the pivot back at the partition where it belongs.
    swap(in_list, partition_index, end)
    return partition_index


def quickselect(in_list, order):
    """Return the order-th order statistic of in_list.
    Note: This function modifies in_list.
    """
    assert 1 <= order <= len(in_list)

    begin = 0
    end = len(in_list) - 1

    while begin <= end:
        pivot_index = random.randint(begin, end)
        partition_index = partition(in_list, begin, end, pivot_index)

        if partition_index > order - 1:
            # The number we want is smaller than the pivot.
            end = partition_index - 1

        elif partition_index < order - 1:
            # The number we want is larger than the pivot.
            start = partition_index + 1

        else:
            # partition_index == order -1, meaning we found the number we want.
            return in_list[partition_index]


class TestQuickSort(unittest.TestCase):
    def test_singleton(self):
        self.assertEqual(1, quickselect([1], 1))

    def test_doubleton(self):
        self.assertEqual(1, quickselect([1, 2], 1))
        self.assertEqual(2, quickselect([1, 2], 2))

    def test_tripleton(self):
        self.assertEqual(-100.5, quickselect([2, -100.5, 12], 1))
        self.assertEqual(2, quickselect([2, -100.5, 12], 2))
        self.assertEqual(12, quickselect([2, -100.5, 12], 3))

    def test_large_list(self):
        for index in range(1, 100):
            self.assertEqual(index, quickselect(list(range(99, 0, -1)), index))

if __name__ == '__main__':
    unittest.main()
